### PR TITLE
Stack recored:Add maximum stack statistics when the task is running

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
Add maximum stack statistics when the task is running
## Impact
procfs
## Testing
use demo test

To put it simply, this function uses gcc's instrumentation function to count the maximum stack usage of each task at runtime, which is very helpful for development.
For more information please read the documentation in the patch

### CODE REVIEW

File: arch/arm/include/irq.h [hunk 1]
```diff
@@ -116,7 +116,7 @@ EXTERN volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];
  ****************************************************************************/
 
 #ifdef CONFIG_SMP
-int up_cpu_index(void);
+int up_cpu_index(void) noinstrument_function ;
```
Comment (pkarashchenko) [inline arch/arm/include/irq.h:119]:
```suggestion
int up_cpu_index(void) noinstrument_function;
```

File: arch/xtensa/src/esp32s3/esp32s3_start.c [hunk 2]
```diff
@@ -170,7 +170,7 @@ uint32_t g_idlestack[IDLETHREAD_STACKWORDS]
  *
  ****************************************************************************/
 
-static void IRAM_ATTR configure_cpu_caches(void)
+noinstrument_function static  void IRAM_ATTR configure_cpu_caches(void)
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_start.c:173]:
```suggestion
noinstrument_function static void IRAM_ATTR configure_cpu_caches(void)
```

File: include/nuttx/instrument.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+typedef CODE void (instrument_fun_t)(FAR void *this_fn,
+                                     FAR void *call_site,
+                                     FAR void *arg);
+
+struct instrument_s
+{
+  sq_entry_t entry;
+  instrument_fun_t *enter;
+  instrument_fun_t *leave;
```
Comment (pkarashchenko) [inline include/nuttx/instrument.h:43]:
```suggestion
  FAR instrument_fun_t *enter;
  FAR instrument_fun_t *leave;
```

File: sched/instrument/instrument.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Does not return.
+ *
+ ****************************************************************************/
+
+void instrument_initialize(void);
+#endif
```
Comment (pkarashchenko) [inline sched/instrument/instrument.h:43]:
```suggestion
void instrument_initialize(void);

#endif
```

**Review by pkarashchenko — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — COMMENTED** [review]
please add commit messages documenting the change

**Pull request merged**